subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/TaskReferenceCodec.kt

```diff
@@ -28,28 +30,44 @@ internal
 object TaskReferenceCodec : Codec<Task> {
 
     override suspend fun WriteContext.encode(value: Task) {
-        if (value === isolate.owner.delegate) {
+        val owner = isolate.owner
+        val delegate = owner.delegate
+
+        if (value === delegate) {
             writeBoolean(true)
         } else {
-            logUnsupported(
-                "serialize",
-                Task::class,
-                value.javaClass,
-                documentationSection = RequirementsTaskAccess
-            )
             writeBoolean(false)
+            val isTaskReferencesAllowed = owner is IsolateOwner.OwnerTask && owner.allowTaskReferences
+            val isTaskFromSameProject = delegate is Task && delegate.project == value.project
+
+            if (isTaskReferencesAllowed && isTaskFromSameProject) {
+                writeNullableString(value.name)
+            } else {
+                logUnsupported(
+                    "serialize",
+                    Task::class,
+                    value.javaClass,
+                    documentationSection = RequirementsTaskAccess
+                )
+                writeNullableString(null)
+            }
         }
     }
 
     override suspend fun ReadContext.decode(): Task? =
         if (readBoolean()) {
             isolate.owner.delegate as Task
         } else {
-            logUnsupported(
-                "deserialize",
-                Task::class,
-                documentationSection = RequirementsTaskAccess
-            )
-            null
+            val taskName = readNullableString()
```

Comment:
🔧 Now that we have 3 possible cases it would be worth defining an enum instead of relying on the  boolean + nullable string.